scylla/tests/integration/tablets.rs

```diff
@@ -0,0 +1,586 @@
+                .build();
+            stmt.set_execution_profile_handle(Some(execution_profile.into_handle()));
+
+            async move { session.execute(&stmt, values_ref).await }
+        })
+    });
+
+    try_join_all(tasks).await
+}
+
+async fn send_unprepared_query_everywhere(
+    session: &Session,
+    cluster: &ClusterData,
+    query: &Query,
+) -> Result<Vec<QueryResult>, QueryError> {
+    let tasks = cluster.get_nodes_info().iter().flat_map(|node| {
+        let shard_count: u16 = node.sharder().unwrap().nr_shards.into();
+        (0..shard_count).map(|shard| {
+            let mut stmt = query.clone();
+            let policy = SingleTargetLBP {
+                target: (node.clone(), Some(shard as u32)),
+            };
+            let execution_profile = ExecutionProfile::builder()
+                .load_balancing_policy(Arc::new(policy))
+                .build();
+            stmt.set_execution_profile_handle(Some(execution_profile.into_handle()));
+
+            async move { session.query(stmt, &()).await }
+        })
+    });
+
+    try_join_all(tasks).await
+}
+
+fn frame_has_tablet_feedback(frame: ResponseFrame) -> bool {
+    let response =
+        scylla_cql::frame::parse_response_body_extensions(frame.params.flags, None, frame.body)
+            .unwrap();
+    match response.custom_payload {
+        Some(map) => map.contains_key("tablets-routing-v1"),
+        None => false,
+    }
+}
+
+fn count_tablet_feedbacks(
+    rx: &mut mpsc::UnboundedReceiver<(ResponseFrame, Option<TargetShard>)>,
+) -> usize {
+    std::iter::from_fn(|| rx.try_recv().ok())
+        .map(|(frame, _shard)| frame_has_tablet_feedback(frame))
+        .filter(|b| *b)
+        .count()
+}
+
+async fn prepare_schema(session: &Session, ks: &str, table: &str, tablet_count: usize) {
+    session
+        .query(
+            format!(
+                "CREATE KEYSPACE IF NOT EXISTS {} 
+            WITH REPLICATION = {{'class' : 'NetworkTopologyStrategy', 'replication_factor' : 2}}
+            AND tablets = {{ 'initial': {} }}",
+                ks, tablet_count
+            ),
+            &[],
+        )
+        .await
+        .unwrap();
+    session
+        .query(
+            format!(
+                "CREATE TABLE IF NOT EXISTS {}.{} (a int, b int, c text, primary key (a, b))",
+                ks, table
+            ),
+            &[],
+        )
+        .await
+        .unwrap();
+}
+
+/// Tests that, when using DefaultPolicy with TokenAwareness and querying table
+/// that uses tablets:
+/// 1. When querying data that belongs to tablet we didn't receive yet we will
+///    receive this tablet data at some point.
+/// 2. When we have all tablets info locally then we'll never receive tablet info.
+///
+/// The test first sends 100 queries per tablet and expects to receive tablet info.
+/// After that we know we have all the info. The test sends the statements again
+/// and expects to not receive any tablet info.
```

Comment:
You seem to use both "tablet info" and "tablet data" terms. Are they the same?

Reply:
Yes, I'll stick to info (or just tablet)